### PR TITLE
No bootstrap

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

**Comment by sschuberth** [comment]
FYI @oheger-bosch.

File: cli/src/main/kotlin/commands/RequirementsCommand.kt
Comment (fviernau) [inline cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: commit: Our internal setup still depends on that mechanism, so we'd need to fix this internally before this PR gets merged.

Comment (sschuberth) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: Oh, really? You're bootstrapping ScanCode on every ORT pipeline run? If so, that's also an opportunity to save a few minutes (as I understand you're working on general performance improvements), esp. if you do `scancode --reindex-licenses` beforehand, like we do in the official `Dockerfile`.

Comment (sschuberth) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: If nothing else is blocking the merge, then I'll leave the merge to you, so you can do it when time suits you.

Comment (sschuberth) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: PS: After this PR I was also considering to upgrade everything to ScanCode 3.2.3 final.

Comment (fviernau) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: Thanks @sschuberth . I'm aware of the optimization potential, have spent quite a bit of work on it but ran into some really weird issues which is why we don't have it in place yet.

Comment (sschuberth) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: So, does this mean we'll not be able to merge this PR any time soon?

Comment (fviernau) [reply cli/src/main/kotlin/commands/RequirementsCommand.kt:113]: No it doesn't. It means that I'd have a hard time providing a timeline.

File: scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt
Comment (mnonnenmacher) [inline scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: This can give the false impression that everything is fine if a scanner is not installed. If we want to continue to support those scanners we should install them on CI instead of skipping the tests IMO.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]

File: scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt
Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: I thought about that, but my thinking was that someone who does not have scanner A installed would also not care if a test for scanner A breaks. I realize the that the question you raise is slightly different, though: Should *we* ensure in our tests that the implementation does not break although we're *not* using the implementation ourselves? And I'm a bit undecided about that. Of course it would be good practice to do so, but it would further increase our CI time due to installation time and run time. So shouldn't we maybe just "risk" a breakage here, and rely on potential users to report it?

**Review by tsteenbe — CHANGES_REQUESTED** [review]
If we removing bootstrapping ScanCode, I propose you update the README's Scanner section and https://github.com/oss-review-toolkit/ort/blob/master/docs/getting-started.md#5-run-the-scanner to mention that users have to install ScanCode locally or use use provided DockerFile to build their own ORT Docker image.

File: scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt
Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: Could you please comment on this again @mnonnenmacher? I'd like to avoid adding all scanners to CI, but still remove the bootstrapping.

Comment (fviernau) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: How about adding a test dedicated for verifying that the scanner is installed?

If you have that, then you could silently skip all other scanner dependent tests in case the scanner is not installed.

Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: You mean, instead of adding `enabledIf = { scanner.isInPath() }` to each test case, have a dedicated test case for that, and then somehow enable / disable the other test cases? Isn't the current solution a more readable way to basically do the same?

Comment (fviernau) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: No, I meant in addition to `enabledIf = { scanner.isInPath() }`.

Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: Then I'm not getting what the purpose of such a test case would be. Is your aim to make it visible in the test results when a scanner is not installed, by making a test fail in that case? But then we'd have failing tests in our CI as it does not have (most of) the scanners installed. Sorry, I must be missing something, but I really do not get it.

Also, how would you enable / disable the other test cases concretely? Of course we could add some test-suite-global boolean flag for a check, but that seems less elegant to me than just relying on the `enabledIf = { scanner.isInPath() }` mechanism.

Comment (fviernau) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: > Is your aim to make it visible in the test results when a scanner is not installed, by making a test fail in that case? 

yes that was the idea.

> But then we'd have failing tests in our CI as it does not have (most of) the scanners installed. Sorry, I must be missing something, but I really do not get it.

I wasn't aware of that. I falsely assumed that all our tests would actually be enabled in our CI. 

> Also, how would you enable / disable the other test cases concretely? 

With `enabledIf = { scanner.isInPath() }`.

> Of course we could add some test-suite-global boolean flag for a check, but that seems less elegant...

That was a misunderstanding, I didn't mean to propose that.

Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: > I falsely assumed that all our tests would actually be enabled in our CI.

Well *these* tests we talk about here *are* currently enabled in CI. But they only work because of the bootstrapping feature, and they would start to fail once the bootstrapping feature is removed, as the scanners are not installed as part of the CI scripts.

Comment (fviernau) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: > Well these tests we talk about here are currently enabled in CI. But they only work because of the bootstrapping feature, and they would start to fail once the bootstrapping feature is removed, as the scanners are not installed as part of the CI scripts.

Similar here I also made a false assumption: I assumed removing the bootstrapping implies installing all scanners so that all tests are executed on CI. 

If these assumption do not hold, then my idea does not make sense - I agree with that.

Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: > I assumed removing the bootstrapping implies installing all scanners so that all tests are executed on CI.

As mentioned [here](https://github.com/oss-review-toolkit/ort/pull/3254#discussion_r526806445), I'd like to avoid adding all scanners to all CI scripts, and my concrete question to @mnonnenmacher and you is: Could you live with that and only make the test cases execute if a scanner is present, or would you insist on always testing all scanners that we theoretically support?

Comment (fviernau) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: I prefer having the tests for all supported scanners enabled, to have some confidence that the integration actually works. Not necessarily on (every) PSV run but at least on a regular basis or if something relevant changed, but that's a different topic.

Anyhow, I won't insist on it. I find it more important to spend the time on the more general topic / plan for reducing the PSV feedback time. So I'll leave the decision up to you.

Comment (mnonnenmacher) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: If we list these scanners in the readme as supported scanners (currently the README says "reference implementations") they should be tested. Maybe only nightly if it is slow. Otherwise we should explicitly mark them as unstable somehow, and maybe also keep only one of them as a show-case. I understand the main purpose of this PR is get rid of the bootstrap code to reduce the amount of code we need to maintain, then why not also get rid of the scanner implementations that no one uses anyway? And only ScanCode is installed in the Docker image (if I'm right), so the other scanners would not work out-of-the box with bootstrapping removed.

Comment (sschuberth) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: Hmm. Maybe we could move out the "reference implementations" to separate projects in a separate repository to show-case the plugin mechanism at the same time? Which is something that you were planning to do in the old *plugin-example* branch anyway, right?

Comment (mnonnenmacher) [reply scanner/src/funTest/kotlin/scanners/AbstractScannerFunTest.kt:79]: Good idea, what about changing "reference implementations" to something like "demo implementations" now to emphasize they are not as stable as the ScanCode implementation and moving them out when we start splitting out also other things like spdx-utils and the clients?

**Comment by sschuberth** [comment]
Abandoning this PR in favor of keeping the bootstrapping, but only in case a particular scanner is not found at all (also see #4669).

**Pull request closed without merging**
